modules/core/arrow-core/src/main/kotlin/arrow/core/extensions/listk.kt

```diff
@@ -279,3 +285,25 @@ interface ListKAlternative : Alternative<ForListK>, ListKApplicative {
   override fun <A> Kind<ForListK, A>.orElse(b: Kind<ForListK, A>): Kind<ForListK, A> =
     (this.fix() + b.fix()).k()
 }
+
+@extension
+interface ListKSemialign : Semialign<ForListK>, ListKFunctor {
+  override fun <A, B> align(
+    left: Kind<ForListK, A>,
+    right: Kind<ForListK, B>
+  ): Kind<ForListK, Ior<A, B>> {
+    val l = left.fix()
+    val r = right.fix()
+
+    fun <T> maybeGet(list: List<T>, idx: Int) =
+      if (idx >= list.size) Option.empty() else list[idx].some()
+
+    val maxSize = max(l.size, r.size)
+
+    return (0..maxSize).map {
+      maybeGet(l, it) toT maybeGet(r, it)
+    }.filterMap {
+      Ior.fromOptions(it.a, it.b)
+    }.k()
```

Comment:
I don't dislike this (especially the use of fromOptions is nice^^), but is there any reason to not write it similar to the haskell one?
So that would be:
```kotlin
fun align(ls: List<Int>, rs: List<Int>): List<Ior<Int, Int>> = when {
    ls.isEmpty() -> rs.map { it.rightIor() }
    rs.isEmpty() -> ls.map { it.leftIor() }
    else -> listOf(Ior.Both(ls.first(), rs.first())) + align(ls.tail(), rs.tail())
}
```
If there is any reason for it then leave it, up to you.